### PR TITLE
SY-4342: Fix camelToSnake Single-Letter Round-Trip and Improve Performance

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-4342](https://linear.app/synnax/issue/SY-4342)

## Description

`camelToSnake` was not a true inverse of `snakeToCamel` for identifiers containing a single-letter segment. `setXChannel` encoded to `set_xChannel` instead of `set_x_channel` because the old regex `/([a-z0-9])([A-Z])/g` only split a lowercase-to-uppercase boundary and never split the `XC` uppercase-to-uppercase boundary before `Channel`. As a result, dispatching internally-tagged union actions whose discriminator has a single-letter token (e.g. the line plot `set_x_channel` action) failed server-side with `variant "set_x_channel": missing payload for declared variant`, because the payload key never matched the Go struct's `set_x_channel` json tag. Actions without single-letter segments (`set_line_color`, `add_channel`) were unaffected, which is why only x-axis channel edits broke.

The fix treats an interior capital that follows another capital but precedes a lowercase letter as a word break, so the round-trip holds (`set_x_channel` ⇄ `setXChannel`). A fully-uppercase semantic token with no trailing lowercase (e.g. `NS=1;ID=5`) is left untouched.

Since this converter runs on every key of every outgoing payload, the implementation also drops the regex in favor of a char-code scan that returns the input unchanged when nothing needs converting and only allocates once a word break is found. It benchmarks ~2x faster than the previous regex on realistic short wire keys (and faster across the existing `camelToSnake` benchmark suite), with no behavior change for any previously-correct input.

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR fixes `camelToSnake` so it correctly handles single-letter camelCase segments (e.g. `setXChannel → set_x_channel`), resolving a server-side action dispatch failure for x-axis channel edits. It also replaces the regex implementation with a char-code scan that short-circuits on already-snake strings, avoiding per-match allocations.

- **Bug fix**: adds an uppercase-uppercase-lowercase boundary rule so `XC` in `setXChannel` is split correctly, making `camelToSnake` a true inverse of `snakeToCamel` for the single-letter-followed-by-multi-letter pattern.
- **Performance**: the new char-code loop returns the input string unchanged when no word break is found, eliminating regex engine overhead on the hot path (every key of every outgoing payload).
- **Tests**: two new table entries and a round-trip test cover the primary fixed case, but the consecutive-single-letter-segment edge case (`foo_x_y → fooXY → foo_xY`) remains untested and broken.

<h3>Confidence Score: 4/5</h3>

Safe to merge for the stated use case; the fix resolves the x-axis channel dispatch bug and does not regress any previously-correct input.

The core fix is correct and well-tested for its primary scenario. The only gap is that identifiers with two or more consecutive single-letter segments (e.g. foo_x_y) still fail the round-trip, contrary to the broad claim in the code comments. This is a narrow edge case absent from the current API surface, but the invariant is weaker than advertised.

x/ts/src/caseconv/caseconv.ts — the breaksSnakeWord predicate and its interaction with trailing or consecutive uppercase single-letter segments deserves a second look.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| x/ts/src/caseconv/caseconv.ts | Replaces the regex-based camelToSnakeStr with a char-code scan; correctly fixes the XC uppercase-uppercase-lowercase boundary case (setXChannel → set_x_channel), but the round-trip invariant still breaks for identifiers with two consecutive single-letter segments (foo_x_y → fooXY → foo_xY). |
| x/ts/src/caseconv/caseconv.spec.ts | Adds two new string table entries and a round-trip test for set_x_channel; the new tests cover the primary fixed case but leave the consecutive-single-letter-segment edge case (foo_x_y) untested. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["camelToSnakeStr(str)"] --> B["Scan from i=1, find first breaksSnakeWord"]
    B --> C{Any break found?}
    C -- No --> D["Return str unchanged (no allocation)"]
    C -- Yes --> E["Build res = slice(0,i) + '_' + lower(str[i])"]
    E --> F["For each remaining char i++"]
    F --> G{breaksSnakeWord at i?}
    G -- Yes --> H["Append '_' + lower(str[i])"]
    G -- No --> I["Append str[i] verbatim"]
    H --> F
    I --> F
    F -- done --> J["Return res"]
    subgraph breaksSnakeWord
        K["str[i] is A-Z?"] -- No --> L["false"]
        K -- Yes --> M["prev is a-z or 0-9?"]
        M -- Yes --> N["true (standard camel boundary)"]
        M -- No --> O["prev is A-Z AND next is a-z?"]
        O -- Yes --> P["true (XC interior capital)"]
        O -- No --> Q["false (trailing cap / semantic token)"]
    end
```

<sub>Reviews (1): Last reviewed commit: ["SY-4342: Fix camelToSnake single-letter ..."](https://github.com/synnaxlabs/synnax/commit/4ea549aefcff7bc5e2e1fbbd1b1bb2c0da513482) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36245381)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

**Review by pjdotson — CHANGES_REQUESTED** [review]
approved after Greptile comment addressed

File: x/ts/src/caseconv/caseconv.ts [hunk 1]
```diff
@@ -244,13 +244,38 @@ const createConverter = (
...
+  let i = 1;
+  while (i < n && !breaksSnakeWord(str, i)) i++;
+  if (i >= n) return str;
+  let res = `${str.slice(0, i)}_${String.fromCharCode(str.charCodeAt(i) + 32)}`;
+  for (i++; i < n; i++)
+    res += breaksSnakeWord(str, i)
+      ? `_${String.fromCharCode(str.charCodeAt(i) + 32)}`
+      : str[i];
+  return res;
+};
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/caseconv/caseconv.ts]: Good catch, addressed. `camelToSnake` now tracks whether the current uppercase run was entered from a lowercase/digit, so every capital in such a run breaks a word: `fooXY` (from `foo_x_y`) and `setXYChannel` (from `set_x_y_channel`) round-trip correctly, while a leading run like `NS=1;ID=5` is still left untouched. Added `fooXY`/`setXYChannel` to the conversion table and extended the round-trip test to `foo_x_y` and `set_x_y_channel`. The prefix scan is preserved so already-snake keys still allocate nothing (no perf regression).
